process/caller.py

```diff
@@ -1,25 +1,172 @@
+        """Evaluate models until results are idempotent, then write output files.
+
+        Ensure all outputs in mfile are idempotent before returning, by
+        evaluating models multiple times. Typically used at the end of an
+        optimisation, or in a non-optimising evaluation. Writes OUT.DAT and
+        MFILE.DAT with final results.
+
+        :param xc: optimisation parameter
+        :type xc: np.ndarray
+        :param ifail: return code of solver
+        :type ifail: int
+        :raises RuntimeError: if values are non-idempotent after successive
+        evaluations
+        """
+        # TODO The only way to ensure idempotence in all outputs is by comparing
+        # mfiles at this stage
+        previous_mfile_arr = None
+        # Path of output file prefix required for writing intermediate idempotence-
+        # checking mfiles in the same directory as input file
+        output_prefix = ft.global_variables.output_prefix
+
+        # Evaluate models up to 10 times; any more implies non-converging values
+        for _ in range(10):
+            # Divert OUT.DAT and MFILE.DAT output to scratch files for
+            # idempotence checking
+            ft.init_module.open_idempotence_files()
+            self._call_models_once(xc)
+            # Write mfile
+            finalise(self.models, ifail)
+
+            # Extract mfile data
+            mfile_path = f2py_compatible_to_string(output_prefix) + "IDEM_MFILE.DAT"
+            mfile = MFile(mfile_path)
+            mfile_data = {}
+            for var in mfile.data.keys():
+                mfile_data[var] = mfile.data[var].get_scan(-1)
+
+            # Extract floats from mfile dict into array for straightforward
+            # comparison: only compare floats
+            current_mfile_arr = np.array(
+                [val for val in mfile_data.values() if type(val) is float]
+            )
+            if previous_mfile_arr is None:
+                # First run: need another run to compare with
+                logger.debug(
+                    "New mfile created: evaluating models again to check idempotence"
+                )
+                previous_mfile_arr = np.copy(current_mfile_arr)
+                continue
+
+            if self.check_agreement(previous_mfile_arr, current_mfile_arr):
+                # Previous and current mfiles agree (idempotent)
+                logger.debug("Mfiles idempotent, returning")
+                # Divert OUT.DAT and MFILE.DAT output back to original files
+                # now idempotence checking complete
+                ft.init_module.close_idempotence_files()
+                # Write final output file and mfile
+                finalise(self.models, ifail)
+                return
+
+            # Mfiles not yet idempotent: re-evaluate models
+            logger.debug("Mfiles not idempotent, evaluating models again")
+            previous_mfile_arr = np.copy(current_mfile_arr)
+
+        raise RuntimeError(
```

Comment:
This error, or an exception raised when calling the models before final idempotence will cause the idempotence check files to be left in the filesystem. Are they useful? Or does an `except` block need to remove them?

Reply:
Good point: they might be useful, but most of the time they'd be an annoyance. I've added an `except` block.